### DISCUSSION TITLE
Tag System

### DISCUSSION THREAD

**Discussion by Hiiragi283** [CleanroomMC/Cleanroom#104] [Draft]
### Summary

Add tag system and replace ore dictionary

### Goals

- Add tag system like modern mc
- Replace ore dict with tag

### Non Goals

None.

### Motivation

- Ore Dictionary is added any time, so different value may be return depending on the timing
- Ore Dictionary has two keys but it's difficult to understand for newbies
- Tag can be applied any classes 

### Description

```java
public final class TagManager<V> {

    public final ImmutableMultiMap<String, V> tagMap;

    public TagManager(MultiMap<String ,V> tagMap) {
        this.tagMap = ImmutableMultiMap.copyOf(tagMap);
    }

    public @NotNull ImmutableSet<V> get(String name) {
        return tagMap.get(name);
    }

    public boolean contains(String name) {
        return tagMap.containsKey(name);
    }

}

public final class TagRegisterEvent<V> extends Event {

    private final MultiMap<String, V> tagMap;

    public TagRegisterEvent(MultiMap<String ,V> tagMap) {
        this.tagMap = tagMap;
    }

    public void add(String name, V value) {
        tagMap.put(name, value);
    }

}
```

### Dependencies

Maybe use `IItemProvider` to store item value.

### References

None.

### Guidelines

- [X] I have and will continue to adhere to the guidelines

**Comment by kappa-maintainer** [comment]
Any classes? How do we fire register event for every class?
And how could we maintain the compatibility to oredict (or not)?
The description need more details about registration and example values too

**Comment by Ecdcaeb** [comment]
For every class is too wide to imple.
Could we just limit it into `IForgeRegistryEntry`?

**Reply by Hiiragi283** [reply]
Yes, but i have no idea how to treat metadata…

**Reply by Ecdcaeb** [reply]
A tag lib for 1.7 here : https://github.com/Team-Starry/PineappleTags/
But need more improvement. And docking with OreDic.

**Reply by Hiiragi283** [reply]
That’s true because mutable tag object is the same as Ore Dictionary and not safety.
My idea is based on vanilla’s builder system but not using JSON
